### PR TITLE
ExcelNPOIStorage: Handle blank fields and allow runtime field removal

### PR DESCRIPTION
**Pull request by marc-lopez**
For scenarios that require supporting files that contain a subset of the record's available fields.  Cell traversal is changed to explicitly set blank values to null as the default behavior skips them, resulting in out-of-index exceptions when matching with the type's fields.

Some interface types were set to public visibility to allow the ExcelNPOIStorage class to perform runtime field removal.

Tackles the same problem as #203

### CODE REVIEW

**Comment by mcavigelli** [comment]
Thank you for your fix!

I only have started to review the code changes. It seems to me that the commits 
contain lots of formatting changes and changed behaviour at the same time.
That makes it hard to review. It is for instance hard to see, if a new interface
is required or if it is just a question of taste.

Commit db3b388 says you have fixed the tests: does the code in Test/Program
show the changed behavior you are trying to achieve? Have you considered
to write as a NUnit test, such as in file ExcelStorageNPOITests in the namespace
FileHelpers.Tests.Tests.DataLink?

Matthias

**Comment by marc-lopez** [comment]
Hello Matthias, The formatting changes is part of the Telerik JustCode's features. They can be rolled back if needed.

The Test/Program fix in db3b388 is limited to silencing missing file exceptions. I can add an additional test for the changed behavior.

Thanks

File: FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs [hunk 1]
```diff
@@ -214,140 +236,174 @@ private string CellAsString(IRow row, int col)
...
+                
                 ICell cell = HSSFCellUtil.GetCell(row, startCol);
-                return new object[] {NPOIUtils.GetCellValue(cell)};
+                return new object[] { NPOIUtils.GetCellValue(cell) };
             }
-            else {
+            else
+            {
+                var values = new List<object>();
                 CellRangeAddress range = new CellRangeAddress(rowNum, rowNum, startCol, startCol + numberOfCols - 1);
```
Comment (mcavigelli) [inline FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs]: Note to self: this is the block with changed behaviour.
Needs unit test.

File: FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs [hunk 2]
```diff
@@ -402,75 +466,91 @@ public override object[] ExtractRecords()
...
+                              ? String.Empty
+                              : values[i].ToString());
             }
-
+            
             return res;
         }
-
+        
+        public void RemoveField(string fieldName)
```
Comment (mcavigelli) [inline FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs]: This is a different change that required the interfaces IRecordInfo and ILineInfo to be public and or extended.
Please move to a different commit.

File: FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs [hunk 3]
```diff
@@ -402,75 +466,91 @@ public override object[] ExtractRecords()
...
+        {
+            base.mRecordInfo.RemoveField(fieldName);
+        }
+        
         private class CellExtractor : ICellHandler
         {
-            private List<object> _cells;
-
+            private readonly List<object> _cells;
+
```
Comment (mcavigelli) [inline FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs]: This enhancement is not related to the other changes.
Please move if possible to an own commit.

File: FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs [hunk 4]
```diff
@@ -402,75 +466,91 @@ public override object[] ExtractRecords()
...
-
+        
         private class CellInserter : ICellHandler
         {
-            private List<object> _cells = null;
-            private List<object>.Enumerator _valuesEnumerator;
-
+            private readonly List<object> _cells = null;
+            private readonly List<object>.Enumerator _valuesEnumerator;
+
```
Comment (mcavigelli) [inline FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs]: See above, please include in that commit.

File: FileHelpers/Core/ILineInfo.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Linq;
+
+namespace FileHelpers
+{
+    public interface ILineInfo
```
Comment (mcavigelli) [inline FileHelpers/Core/ILineInfo.cs]: Please add comments to this interface.
It should be part of the remove feature commit.

File: FileHelpers/Core/IRecordOperations.cs [hunk 6]
```diff
@@ -3,11 +3,12 @@
 
 namespace FileHelpers
 {
-    internal interface IRecordOperations
+    public interface IRecordOperations
```
Comment (mcavigelli) [inline FileHelpers/Core/IRecordOperations.cs]: Please add comments, at least to those members that you have added.
Should be part of the remove feature.

File: FileHelpers/Core/LineInfo.cs [hunk 7]
```diff
@@ -79,23 +79,30 @@ public string Substring(int from, int count)
...
-                return "<EOL>";
-            else
-                return CurrentString;
+            get { return LineString.Substring(mCurrentPos, mLineStr.Length - mCurrentPos); }
         }
-
+        
         /// <summary>
-        /// Extract a single field from the system
+        /// Extract the whole system
```
Comment (mcavigelli) [inline FileHelpers/Core/LineInfo.cs]: Probably not the whole _system_ is extracted, right?

File: FileHelpers/Core/LineInfo.cs [hunk 8]
```diff
@@ -79,23 +79,30 @@ public string Substring(int from, int count)
         #endregion
```
Comment (mcavigelli) [inline FileHelpers/Core/LineInfo.cs]: You have removed a private method. It seems that it is needed for debugging display.
Please revert this change.

File: FileHelpers/Core/LineInfo.cs [hunk 9]
```diff
@@ -246,7 +267,7 @@ public int IndexOf(string foundThis)
         /// </summary>
         /// <remarks>If the input is multi line, this will read next record and remove the original data</remarks>
         /// <param name="line">Line to use</param>
-        internal void ReLoad(string line)
+        public void ReLoad(string line)
```
Comment (mcavigelli) [inline FileHelpers/Core/LineInfo.cs]: I  think by including the interface (for removing fields) other members are now forced to be public as well. That makes it harder to maintain the library and makes it also more complicated to use the library.
Please limit the interface just to those members you really need for removing fields and keep the members internal.

File: FileHelpers/Engines/MultiRecordEngine.cs [hunk 10]
```diff
@@ -1,939 +1,939 @@
...
+                    for (int i = 0; i < mMultiRecordInfo[0].IgnoreFirst && currentLine != null; i++) {
+                        mHeaderText += currentLine + StringHelper.NewLine;
+                        currentLine = freader.ReadNextLine();
+                        mLineNumber++;
+                    }
+                }
+
+                bool byPass = false;
+
+                ILineInfo line = new LineInfo(currentLine) {
```
Comment (mcavigelli) [inline FileHelpers/Engines/MultiRecordEngine.cs]: Please revert the changes in this file. The whole file has changed (line endings).

File: FileHelpers/ErrorHandling/BadUsageException.cs [hunk 11]
```diff
@@ -21,7 +21,7 @@ protected internal BadUsageException(int line, int column, string message)
         /// <summary>Creates an instance of an BadUsageException.</summary>
         /// <param name="message">The exception Message</param>
         /// <param name="line">Line to display in message</param>
-        internal BadUsageException(LineInfo line, string message)
-            : this(line.mReader.LineNumber, line.mCurrentPos, message) {}
+        internal BadUsageException(ILineInfo line, string message)
```
Comment (mcavigelli) [inline FileHelpers/ErrorHandling/BadUsageException.cs]: Change should be part of the remove feature.

File: FileHelpers/Fields/FieldBase.cs [hunk 12]
```diff
@@ -537,39 +537,40 @@ internal string CreateFieldString(object fieldValue)
         /// </summary>
         /// <param name="line">Line handler containing text</param>
         /// <returns></returns>
-        internal object ExtractFieldValue(LineInfo line)
+        internal object ExtractFieldValue(ILineInfo line)
         {
             //-> extract only what I need
+            var castedLine = (LineInfo)line;
```
Comment (mcavigelli) [inline FileHelpers/Fields/FieldBase.cs]: It is suspicious when an implementation has to cast an argument.
It seems, that ILineInfo does not fit well.

Should be part of the remove feature.

File: FileHelpers/FileHelpers.csproj [hunk 13]
```diff
@@ -174,6 +174,8 @@
       <SubType>Code</SubType>
     </Compile>
     <Compile Include="Core\ExtensionsFileHelpers.cs" />
+    <Compile Include="Core\ILineInfo.cs" />
```
Comment (mcavigelli) [inline FileHelpers/FileHelpers.csproj]: These files are part of the remove feature.

File: FileHelpers/Fields/DelimitedField.cs [hunk 14]
```diff
@@ -1,3 +1,4 @@
+using System;
```
Comment (mcavigelli) [inline FileHelpers/Fields/DelimitedField.cs]: Is this using required?

File: FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs [hunk 15]
```diff
@@ -71,48 +81,53 @@ private void InitExcel()
...
-                    mWorkbook = new XSSFWorkbook(file);
-                else
-                    mWorkbook = new HSSFWorkbook(file);
-
+            }
+            
+            using (FileStream file = new FileStream(filename, FileMode.Open, FileAccess.Read))
+            {
+                mWorkbook = WorkbookFactory.Create(file);
+
```
Comment (marc-lopez) [inline FileHelpers.ExcelNPOIStorage/ExcelNPOIStorage.cs]: NPOI reads the file's headers to determine its type. An .xls file formatted as an Excel 2007 file would be read as an .xlsx instead.

Reference: https://github.com/tonyqus/npoi/blob/master/ooxml/SS/UserModel/WorkbookFactory.cs

File: FileHelpers/Fields/FieldBase.cs (hunk 12 above)
Comment (marc-lopez) [inline FileHelpers/Fields/FieldBase.cs]: We can extract the interfaces in a separate PR.

File: FileHelpers/Core/LineInfo.cs (hunk 9 above)
Comment (marc-lopez) [inline FileHelpers/Core/LineInfo.cs]: Agreed. That's one of the reasons why I decided to expose IRecordInfo and ILineInfo are exposed as public: to limit additional maintenance complexity.
